src/modules/aws_ec2/mappers/instance_block_device_mapping.ts

```diff
@@ -0,0 +1,257 @@
+import { EC2, InstanceBlockDeviceMapping as AWSInstanceBlockDeviceMapping } from '@aws-sdk/client-ec2';
+import { Volume as AWSVolume } from '@aws-sdk/client-ec2';
+import { WaiterState } from '@aws-sdk/util-waiter';
+
+import { AwsEc2Module } from '..';
+import { AWS, crudBuilder } from '../../../services/aws_macros';
+import { Context, Crud, IdFields, MapperBase } from '../../interfaces';
+import { Instance, InstanceBlockDeviceMapping, VolumeState } from '../entity';
+
+export class InstanceBlockDeviceMappingMapper extends MapperBase<InstanceBlockDeviceMapping> {
+  module: AwsEc2Module;
+  entity = InstanceBlockDeviceMapping;
+  generateId = (fields: IdFields) => {
+    const requiredFields = ['instanceId', 'volumeId', 'region'];
+    if (
+      Object.keys(fields).length !== requiredFields.length &&
+      !Object.keys(fields).every(fk => requiredFields.includes(fk))
+    ) {
+      throw new Error(`Id generation error. Valid fields to generate id are: ${requiredFields.join(', ')}`);
+    }
+    return `${fields.instanceId}|${fields.volumeId}|${fields.region}`;
+  };
+  entityId = (e: InstanceBlockDeviceMapping) => {
+    return this.module.instanceBlockDeviceMapping.generateId({
+      instanceId: e.instance?.instanceId ?? '',
+      volumeId: e.volume?.volumeId ?? '',
```

Comment:
Why fall back on these to empty string? Are there legitimate `InstanceBlockDeviceMapping` records with this being the case? (If so, how do you distinguish between multiple of them, if, say, a new instance is defined with 3 new volume definitions? There would be 3 of these records that all have the same ID so the engine would think there's only one of them to operate on.)

Reply:
actually the volume is optional here because it actually can be the case in the initial mapping of cleaning volumes with NoDevice. But it will only exist in the database and in the initial mapping definition, not in the cloud